python-threatexchange/threatexchange/extensions/vpdq/vpdq_index.py

```diff
@@ -0,0 +1,100 @@
+    """
+
+    def __init__(
+        self,
+        quality_threshold: int = VPDQ_QUALITY_THRESHOLD,
+    ) -> None:
+        super().__init__()
+        self.index: VPDQHashIndex = VPDQHashIndex()
+        self._entry_idx_to_features_and_entires: t.List[
+            t.Tuple[t.List[vpdq.VpdqFeature], IndexT]
+        ] = []
+        self._index_idx_to_vpdqHex_and_entry: t.List[t.Tuple[int, t.List[int]]] = []
+        self._unique_vpdqHex_to_idx: t.Dict[str, int] = {}
+        self.quality_threshold = quality_threshold
+
+    def add(self, signal_str: str, entry: IndexT) -> None:
+        entry_id = len(self._entry_idx_to_features_and_entires)
+        features = prepare_vpdq_feature(signal_str, self.quality_threshold)
+        self._entry_idx_to_features_and_entires.append((features, entry))
+        # Use hex to represent the feature because it saves the space
+        unique_features = []
+        for f in features:
+            idx = self._unique_vpdqHex_to_idx.get(f.hex)
+            if idx is None:
+                idx = len(self._unique_vpdqHex_to_idx)
+                self._unique_vpdqHex_to_idx[f.hex] = idx
+                self._index_idx_to_vpdqHex_and_entry.append((f.hex, list()))
+                unique_features.append(f)
+            self._index_idx_to_vpdqHex_and_entry[idx][1].append(entry_id)
+        if unique_features:
+            self.index.add_single_video(unique_features)
+
+    def query(self, query_hash: str) -> t.List[IndexMatch[IndexT]]:
+        """Searches this VPDQ index for query hashes within the index that are no more than the threshold away
+        from the query hashes by hamming distance.
+
+        Args:
+            query_hash : Query VPDQ hash
+
+        Returns:
+            List of VPDQIndexMatch
+        """
+        features = prepare_vpdq_feature(query_hash, self.quality_threshold)
+        results = self.index.search_with_distance_in_result(
+            features, VPDQ_DISTANCE_THRESHOLD
+        )
+        query_matched: t.Dict[int, t.Set[str]] = {}
+        index_matched: t.Dict[int, t.Set[int]] = {}
+        matches: t.List[IndexMatch[IndexT]] = []
+        for hash in results:
+            for match in results[hash]:
+                # query_str =>  (matched_idx, distance)
+                vpdq_match, entry_list = self._index_idx_to_vpdqHex_and_entry[match[0]]
+                for entry_id in entry_list:
+                    if entry_id not in query_matched:
+                        query_matched[entry_id] = set()
+                    query_matched[entry_id].add(hash)
+
+                    if entry_id not in index_matched:
+                        index_matched[entry_id] = set()
+                    index_matched[entry_id].add(vpdq_match)
+                # Dist(-1) is meaningless here, because VPDQ match does not return a single dist
+        for entry_id in query_matched.keys():
+            matches.append(
+                VPDQIndexMatch(
+                    -1,
```

Comment:
Suggest using the match percentage (int, 0-100) as the placeholder here, instead of -1, which is a bit misleading.

I agree with you that the types are mismatched here. I would like to redo this to make the distance type an internal detail